File: pkg-r/tests/testthat/apps/tool-weather/app-03-tool-result-simple.R
library(shiny)
library(bslib)
library(ellmer)
library(shinychat)
library(weathR)

get_weather_forecast <- tool(
  function(lat, lon, location_name) {
    forecast <- point_tomorrow(lat, lon, short = FALSE)

    icon <- if (any(forecast$temp > 70)) {
      bsicons::bs_icon("sun-fill")
    } else if (any(forecast$temp < 45)) {
      bsicons::bs_icon("snow")
    } else {
      bsicons::bs_icon("cloud-sun-fill")
    }

    ContentToolResult(
      forecast,
      extra = list(
        display = list(
          title = paste("Weather Forecast for", location_name),
          icon = icon
        )
      )
    )
  },
  name = "get_weather_forecast",
  description = "Get the weather forecast for a location.",
  arguments = list(
    lat = type_number("Latitude"),
    lon = type_number("Longitude"),
    location_name = type_string("Name of the location for display to the user")
  ),
  annotations = tool_annotations(
    title = "Weather Forecast",
    icon = bsicons::bs_icon("cloud-sun")
  )
)

ui <- function(req) {
  page_fillable(
    chat_mod_ui("chat")
  )
}

server <- function(input, output, session) {
  client <<- ellmer::chat("openai/gpt-4.1-nano")
  # client <<- ellmer::chat_ollama(model = "mistral-nemo")
  client$register_tool(get_weather_forecast)
  chat_mod_server("chat", client)
}

shinyApp(ui, server, enableBookmarking = "url")

File: pkg-py/tests/playwright/tools/weather/app_03_tool_result_simple.py
import faicons
from chatlas import ChatOpenAI, ContentToolResult
from shiny.express import app_opts, ui

from shinychat.express import Chat

from . import tools


def get_weather_forecast(
    lat: float, lon: float, location_name: str
) -> ContentToolResult:
    """Get the weather forecast for a location."""
    forecast_data = tools.get_weather_forecast(lat, lon)

    # Determine icon based on temperature
    if forecast_data["temperature_2m"] > 21:
        icon = "sun"
    elif forecast_data["temperature_2m"] < 7:
        icon = "snowflake"
    else:
        icon = "cloud-sun"

    # Return ContentToolResult with extra display metadata
    return ContentToolResult(
        value=forecast_data,
        extra={
            "display": {
                "title": f"Weather Forecast for {location_name}",
                "icon": faicons.icon_svg(icon),
            }
        },
    )


# Create chat client and register tool
chat_client = ChatOpenAI(model="gpt-4.1-nano")
chat_client.register_tool(
    get_weather_forecast,
    annotations={"title": "Weather Forecast"},
)


# The Shiny app
ui.page_opts(title="Weather Tool - Tool Result Simple")
app_opts(bookmark_store="url")


chat = Chat(id="chat")
chat.ui()

chat.enable_bookmarking(chat_client)


@chat.on_user_submit
async def handle_user_input(user_input: str):
    response = await chat_client.stream_async(user_input, content="all")
    await chat.append_message_stream(response)
